Add ClassLink Oauth2 support

nicklathe opened the pull request:
This PR adds ClassLink OAuth2 support via OmniAuth. It does so by adding the following:

- Classlink OmniAuth gem
- Controller action and helper methods in omni_auth_controller.rb
- UI changes to propagate and pre-fill new user signup form
- Adds ClassLink to Manage Linked Accounts UI
- Adds "Continue with ClassLink" button
- Hides ClassLink functionality behind DCDO flag
- Unit tests

The supported flow for OAuth for Code.org from ClassLink will be initiated by clicking on the Code.org icon from within ClassLink.

## Additional UI details

Currently, the Finish Student Account page does not include propagated values to pre-fill data in the form. This _is_ done on the Finish Teacher Account page. I'm not sure if it was a miss or regression that we don't/stopped pre-filling values on the Student form. This PR adds some of those values that would be included from OmniAuth sign-ups. Pre-filling the sign-up form with data we get back from ClassLink is a requirement to get certified with ClassLink.

## Screenshots

**ClassLink initiated OAuth request with Student user, populates `name` and `state` with ClassLink auth data**
<img width="2279" height="1007" alt="Screenshot 2026-01-13 at 2 38 55 PM" src="https://github.com/user-attachments/assets/8f7bedef-46b2-4524-a3b2-32187b3c2d7f" />

**ClassLink initiated OAuth request with Teacher user, populates `first name` and `last name`**
<img width="2282" height="1285" alt="Screenshot 2026-01-13 at 2 40 40 PM" src="https://github.com/user-attachments/assets/4f9c7264-76d9-43bf-88ae-5967b79c9f9b" />

**ClassLink account linking via Manage Linked Accounts**
<img width="1109" height="597" alt="Screenshot 2026-01-14 at 11 22 56 AM" src="https://github.com/user-attachments/assets/f49c1c71-d7b3-42ad-832f-becb5712e9f7" />

**Continue with ClassLink button added to sign_in page**
<img width="2271" height="1189" alt="Screenshot 2026-01-15 at 8 56 34 AM" src="https://github.com/user-attachments/assets/83f4749f-6070-4575-bec0-8bcf12cf827b" />

<!--
  A summary of the change, including any relevant background, motivation, and context.
  If relevant, include a description, screenshots, and/or video of the existing and new behavior.
-->



## Links
<!--  Links to relevant external resources.
      - spec docs, Jira tickets, related PRs, Honeybadger errors, etc.
-->

- Jira: https://codedotorg.atlassian.net/browse/P20-1757

## Testing story
<!--  Does your change include appropriate tests?
      - If so, please describe how the tests included in this PR are sufficient.
      - If not, please explain why this change does not need to be tested. -->



## Deployment strategy
**IMPORTANT:** Push ClassLink creds to Secrets Manager before merging



## Follow-up work
<!--  List any clean-up or technical debt that will be addressed in future work.
      - Include Jira links where applicable.
      - Are there any known limitations or improvements planned? -->



## Privacy
<!--  How does this change impact Student & Teacher privacy?
      - Does this collect, use, share, or modify PII, either new or existing? -->



## Security
<!--  How does this change impact our security surface-area?
      - Do API's touched have the right auth?
      - Do infra changes impact our attack surface or encryption?
-->



## Caching
<!--  How does this change impact caching behavior?
      - Are cache keys or invalidation strategies affected?
      - Will this require cache warming or invalidation after deployment? -->



## PR Creation Checklist:
<!--
  The final step! Before you create your PR, double-check that everything is in order.
  Change [ ] to [X] during creation to check boxes.
-->

- [ ] Tests provide adequate coverage
- [ ] Privacy impacts have been documented
- [ ] Security impacts have been documented
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

nicklathe committed:
Add ClassLink Oauth2 support

  - Add classlink gem
  - Add sign-in button
  - Add callback handler to omniauth controller
  - Propogate valeus to the frontend to pre-fill form values

Signed-off-by: Nick Lathe <nick.lathe@code.org>

nicklathe committed:
Classlink initiated launch changes

  - Remove Classlink login button changes
  - Update Classlink Omniauth initialization to ignore state

nicklathe committed:
Restore missed changes in application.scss

nicklathe committed:
Add unit tests and clean up some logic

nicklathe committed:
Copilot feedback changes

nicklathe committed:
Remove unessary k/v in en.yml

nicklathe committed:
Add Manage Linked Accounts functionality

nicklathe committed:
Fix unit tests and remove unused prop

carl-codeorg commented on dashboard/app/models/authentication_option.rb:
```diff
@@ -47,6 +47,7 @@ class AuthenticationOption < ApplicationRecord
     QWIKLABS = 'lti_lti_prod_kids.qwikcamps.com',
     TWITTER = 'twitter',
     MICROSOFT = 'microsoft_v2_auth',
+    CLASSLINK = 'classlink'
```
This should also be added to `UNTRUSTED_EMAIL_CREDENTIAL_TYPES` right?

nicklathe replied:
Done.

carl-codeorg reviewed (approved):
LGTM, just had one comment. Handling LTI account linking/button login offline.

nicklathe committed:
Add "Continue with ClassLink" button

carl-codeorg commented on dashboard/app/views/devise/shared/_oauth_links.haml:
```diff
@@ -21,7 +21,8 @@
         consult the "Developer OAuth Setup on Localhost" doc in the Engineering drive folder
 
-    - [:google_oauth2, :microsoft_v2_auth, :facebook, :clever].each do |provider|
+    - [:google_oauth2, :microsoft_v2_auth, :facebook, :clever, :classlink].each do |provider|
```
Don't we need to hide this button behind DCDO until we get prod certified?

nicklathe replied:
Oh yeah.

carl-codeorg reviewed (commented):
The logo on the new button is showing up kind of low-res/pixelated for me - any way we can improve it, or would we need an SVG from ClassLink for that?

Also, as a follow-up item we need to add the same button to the LTI account linking page [here](https://github.com/code-dot-org/code-dot-org/blob/staging/dashboard/app/views/devise/sessions/_login_lti_account_linking.haml#L26).

nicklathe committed:
Add ClassLink button to other login pages

carl-codeorg commented:
@codex review

carl-codeorg commented:
@codex review

nicklathe committed:
Add DCDO flag and change to .svg

  - Add DCDO flag to hide ClassLink in ManageLinkedAccounts
  - Add classlink svg, remove png
  - Update logic to use svg for all icons

nicklathe commented on apps/src/accounts/constants.js:
```diff
@@ -11,6 +11,7 @@ export const SingleSignOnProviders = {
   google: 'google_oauth2',
   microsoft: 'microsoft_v2_auth',
   clever: 'clever',
+  classlink: 'classlink',
```
Added DCDO check to hide ClassLink in ManageLinkedAccounts.

nicklathe committed:
Set DCDO flag in ManageLinkedAccounts tests
